### PR TITLE
add tunnelling and treat collections as CMWs

### LINKED ISSUE #55: Request for clarification: array length and mixed encoding
I'm noodling through the idea of a modular evidence collection daemon that just runs installed binaries in a certain location and assembles a cmw-collection together to send off, and I'm wondering how much interpretation of the output it would need, assuming all the binaries output a legal CMW of either CBOR or JSON encoding.

When JC<> is used, it's only when the CDDL pun breaks down and you need base64 encoding for binary. If you have your cmw-collection as a CBOR map, and we can use, say, the basename of the evidence collection binary to key the result, but the result is just bytes that are either CBOR or JSON, that doesn't seem to fit into the "cmw is either CBOR or JSON".

It seems that certainly a JSON cmw-collection that contains a CBOR-encoded cmw does not parse unless you're allowed to wantonly swap your interpretation of the CDDL as for JSON or CBOR as you see fit (say with the decapsulation algorithm), given that `cmw` doesn't have a base64 encoded string alternate to store a CBOR-encoded cmw.

Can we say that

```
cmw-collection = {
  + cmw-collection-entry-label => cmw
}
```

should be

```
cmw-collection = {
  + cmw-collection-entry-label => cmw-decap
}
cmw-decap = JC<jc-cmw, bytes .cbor cmw>
jc-cmw = cmw / base64-cmw
base64-cmw = base64-string .cbor cmw
```

I'm unsure if the `.cbor` control operator is allowed to be applied to a base64-string technically it's for byte strings and since "one can use CDDL with JSON by limiting oneself to what can be represented in JSON.  Roughly speaking, this means leaving out byte strings"

I don't think we can use `bytes .cbor cmw` without inserting the major type byte 0x04 before the output of any of the evidence modules, so do we just define a `.feature cmw-decap` to apply the decap algorithm to arbitrary bytes instead of this JC stuff?

### PR DESCRIPTION
**Pull request by thomas-fossati**
* merge collection in the top-level CMW production
* rework the X.509 extension (CMWCollection -> CMW)
* remove special-casing collection in all IANA registrations
* add bi-directional tunnelling between JSON and CBOR serialisations
* rework CDDL assembly & test
* add new examples

Fix #55

### CODE REVIEW

**Comment by thomas-fossati** [comment]
cc @deeglaze

File: draft-ietf-rats-msg-wrap.md [hunk 1]
```diff
@@ -153,34 +153,39 @@ document:
...
+Both the `json-array` and `cbor-array` have the same fields except for slight differences in the types discussed below.
 
 ~~~ cddl
 {::include cddl/cmw-array.cddl}
 ~~~
 {: #fig-cddl-array artwork-align="left"
    title="CDDL definition of the Array format"}
 
-It is composed of three members:
+Each contains three members:
```
Comment (deeglaze) [inline draft-ietf-rats-msg-wrap.md:172]: Do you not say "two to three" when one is optional?

File: draft-ietf-rats-msg-wrap.md [hunk 2]
```diff
@@ -234,6 +232,57 @@ A consumer can always distinguish tags that have been derived via
...
+   title="CDDL definition of the CMW collection format"}
+
+Although initially designed for the composite Attester use case, the CMW collection can be repurposed for other use cases requiring CMW aggregation.
+
+## CMW Tunnel {#cmw-tunnel}
+
+The CMW tunnel type ({{fig-cddl-tunnel}}) allows for moving a CMW in one serialization format, either JSON or CBOR, into a collection that uses the opposite serialization format.
+
+Both tunnel types are arrays with two elements.
+The first element, a fixed text string starting with a `#`, acts as a "magic" number.
```
Comment (deeglaze) [inline draft-ietf-rats-msg-wrap.md:263]: Although everything is encoded as numbers on computers, saying that a string starting with `#` is a magic number doesn't sound right to me. In these cases I use the term "sentinel value".

File: draft-ietf-rats-msg-wrap.md [hunk 3]
```diff
@@ -234,6 +232,57 @@ A consumer can always distinguish tags that have been derived via
...
+The `#`, which is not an acceptable start symbol for the `Content-Type` production ({{collected-cddl}}), makes it possible to disambiguate a CMW tunnel from a CMW array.
+
+~~~ cddl
+{::include cddl/cmw-tunnel.cddl}
+~~~
+{: #fig-cddl-tunnel artwork-align="left"
+   title="CDDL definition of the CMW tunnel format"}
+
+The conversion algorithms are described in the following sub sections.
```
Comment (deeglaze) [inline draft-ietf-rats-msg-wrap.md:272]: Inconsistent use of "subsection" and "sub section" within the same doc. I recommend the former.

**Comment by deeglaze** [comment]
Looks good apart from nits. Thanks Thomas :)

**Review by nedmsmith — APPROVED** [review]
LGTM. Some of the word choice suggestions seem reasonable to incorporate. These could be refined later also.

**Comment by nedmsmith** [comment]
After reviewing the cddl naming I think the following would be more intuitive choices:
```
cmw = cmwj / cmwc
cmwj = cmwj-record / cmwj-collection
cmwc = cmwc-record / cmwc-collection / cmw-cbor-tag
cmwj-collection = { + text => cmwj / cmw-c2j-tunnel }
cbor-collection = { + (int / text) => cmwc / cmw-j2c-tunnel }
cmw-c2j-tunnel = [ "#cmw-c2j-tunnel", base64-string ]
cmw-j2c-tunnel = [ "#cmw-j2c-tunnel", bytes ]
cmw-cbor-tag = #6.<0..18446744073709551615>(bytes)
cmwj-record = [
  type: media-type
  value: base64-string
  ? ind: uint .bits cm-type
]
cmwc-record = [
  type: coap-content-format-type / media-type
  value: bytes
  ? ind: uint .bits cm-type
]
```

**Comment by thomas-fossati** [comment]
> After reviewing the cddl naming I think the following would be more intuitive choices:
> 
> ```
> cmw = cmwj / cmwc
> [...]
> ```

The choice of avoiding extra prefixes was on purpose.  With CDDL 2 imports I'd expect modules that depend on cmw to
```
;# import rfcxxxx as cmw
```
and then use it as
```
my = {
  evidence: cmw.json_array
}
```
If we add `cmw` to all the data items we end up with
```
my = {
  evidence: cmw.cmwj
}
```
which stutters.

**Comment by nedmsmith** [comment]
>With CDDL 2 imports

Which version of cddl tooling recognizes CDDL2 `;#` imports?

**Comment by nedmsmith** [comment]
Applying the `;#` imports convention the full schema looks like the following(?):
```
cmw = cmwj / cmwc
cmwj = json-record / json-collection
cmwc = cbor-record / cbor-collection / cbor-tag
json-collection = { + text => cmwj / c2j-tunnel }
cbor-collection = { + (int / text) => cmwc / j2c-tunnel }
c2j-tunnel = [ "#cmw-c2j-tunnel", base64-string ]
j2c-tunnel = [ "#cmw-j2c-tunnel", bytes ]
cbor-tag = #6.<0..18446744073709551615>(bytes)
json-record = [
  type: media-type
  value: base64-string
  ? ind: uint .bits cm-type
]
cbor-record = [
  type: coap-content-format-type / media-type
  value: bytes
  ? ind: uint .bits cm-type
]
```

**Comment by nedmsmith** [comment]
BTW: should there be a tunnel for X.509 encoded CMWs?

**Comment by deeglaze** [comment]
I thought CMWs were only added as an extension to a certificate as added context to understand a signature, not as another way to aggregate CMWs, no?

**Comment by nedmsmith** [comment]
> I thought CMWs were only added as an extension to a certificate as added context to understand a signature, not as another way to aggregate CMWs, no?

A cmw-collection, conceptually, is a representation of a "slice" of an attested system. A certificate path is used by DICE layering to report evidence for a slice that is analogous to at boot sequence (typically). A cmw-collection could be used in the same way to report a slice through a system. The end of the cert path might still be a TE that has evidence, but not represented in a cert, maybe its a CWT / JWT. The collection therefore, is the cert path plus the token / tokens. The CMW extension to X.509 allows the evidence (as a certificate payload) to be cbor/json while continuing to use X.509 signature format. The certificate is just a container, the same way a token is just a container. A cmw-collection can contain either signed or unsigned payloads (the media-type can be used to signal which form). It's possible to include an x.509 collection in a cwt/jwt using the x5bag claim. Oddly, this is added to the protected headers rather than the payload. But semantically, it is a cmw-collection, but not a first class citizen cmw-collection. It might be as simple as defining an xtoc-tunnel as x5bag.

**Comment by deeglaze** [comment]
I don't see that specifically as a CMW but as a media-type for DICE evidence the form of an x.509 certificate path, no? The tunnel is just here because the CDDL permits both JSON and CBOR.

**Comment by nedmsmith** [comment]
> I don't see that specifically as a CMW but as a media-type for DICE evidence the form of an x.509 certificate path, no? The tunnel is just here because the CDDL permits both JSON and CBOR.

A cmw  could contain a PEM certificate i.e., media-type="application/pem-certificate-chain". If the surrounding cmw-collection is in cbor format, it might make sense to convert the pem cert to cbor. The motivation may be similar to the motivation that wants to convert a jwt to cbor.

This is different from the ASN.1 that requires a CMW certificate extension to choose either JSON / CBOR formatting as the end result could be a PEM encoded certificate regardless. 

It is reasonable for attesting environments to produce a certificate as a form of evidence (as in many DICE implementations) or as an endorsement  (as a root-of-trust key typically has a mfg issued certificate). 

The current definition would allow cmw-collection to include a PEM certificate, but wouldn't be able to perform the binary/text, text/binary translations outlined by the tunnelling mechanism. If its relevant to convert JSON (text) to CBOR (binary) then it seems relevant to convert PEM (text) to CBOR (binary). I don't see the need to convert PEM (text) to JSON (text) since both are text based approaches. 

If the cmw-collection processor can handle equally, text and binary representations, then I don't see the value of tunnels.

**Comment by deeglaze** [comment]
What de you mean to convert the pem to cbor? A base64-encoded string is still representable as bytes in the cbor schema. If there's a cbor encoding of certificate chains, wouldn't you just use that media type instead? I feel like I'm missing something.

**Comment by thomas-fossati** [comment]
> > I don't see that specifically as a CMW but as a media-type for DICE evidence the form of an x.509 certificate path, no? The tunnel is just here because the CDDL permits both JSON and CBOR.
> 
> A cmw could contain a PEM certificate i.e., media-type="application/pem-certificate-chain". If the surrounding cmw-collection is in cbor format, it might make sense to convert the pem cert to cbor.

I don't think so.  The PEM cert is just bytes stashed in the `cbor-array.value`.
The `cbor-array.type` will tell the receiver how to interpret it.  There is no reason for doing any format conversion.

> The motivation may be similar to the motivation that wants to convert a jwt to cbor.

Sorry, I don't see the analogy.

> It is reasonable for attesting environments to produce a certificate as a form of evidence (as in many DICE implementations) or as an endorsement (as a root-of-trust key typically has a mfg issued certificate).

Absolutely.  But, as Dionna [pointed out](https://github.com/ietf-rats-wg/draft-ietf-rats-msg-wrap/pull/65#issuecomment-1944946698), in a CMW world, this is handled with media types.

> If the cmw-collection processor can handle equally, text and binary representations, then I don't see the value of tunnels.

The use case for tunnels is when one needs to aggregate already constructed CMWs (rather than raw evidence), and those CMWs have a different serialisation than the one used/implemented by the aggregator.

**Comment by thomas-fossati** [comment]
> > With CDDL 2 imports
> 
> Which version of cddl tooling recognizes CDDL2 `;#` imports?

The latest `cddlc`.

> Are we married to `-array` vs. `-record`?

I am neutral.  It's only some search & replace over the CDDL and the prose which I am too lazy to tackle at the moment :-)

**Comment by nedmsmith** [comment]
> > Are we married to `-array` vs. `-record`?
> I am neutral. It's only some search & replace over the CDDL and the prose which I am too lazy to tackle at the moment :-)

See PR #67

**Comment by thomas-fossati** [comment]
> > > Are we married to `-array` vs. `-record`?
> > > I am neutral. It's only some search & replace over the CDDL and the prose which I am too lazy to tackle at the moment :-)
> 
> See PR #67

Awesome, thanks!  Can we merge this one?

**Comment by nedmsmith** [comment]
> Awesome, thanks! Can we merge this one?

Note: I pulled from tunnels branch which pulled in PR #65 changes. Maybe #65 isn't needed if #67 is merged?

**Comment by thomas-fossati** [comment]
> > Awesome, thanks! Can we merge this one?
> 
> Note: I pulled from tunnels branch which pulled in PR #65 changes. Maybe #65 isn't needed if #67 is merged?

We could close this and move everything there (including the "Fixes" annotations).  Or we could merge this and then you rebase #67 on main.

As you prefer.

**Comment by nedmsmith** [comment]
OK to squash and merge then?

**Comment by thomas-fossati** [comment]
Superseded by #67

**Pull request closed without merging**